### DISCUSSION TITLE
How to Budgets in Snowflake?

### DISCUSSION THREAD

**Discussion by onerishabh** [littleK0i/SnowDDL#148] [General]
Hey, 

I understand that SnowDDL is adding features as quickly as possible from Snowflake. I just wanted to understand if there is anyway that I can deploy budgets on Snowflake using SnowDDL? Maybe a combination of some technical roles or adding new blue prints?

I'm happy to work towards this feature as well if required!

Thanks,

**Comment by littleK0i** [comment]
We did not implement budgets on purspose, since it is a deeply flawed feature which does not address the problem.

The following issues were noted the last time I checked:

- It is not a proper object type, but a class with object instances. Instead of running standard CREATE / ALTER / DROP / GRANT it requires to use database roles and call class instance methods. This is a relatively big departure from lifecycle model currently implemented in resolvers.
- It runs on monthly periods only. Budget object creation time matters.
- There is a limit of only 100 budgets per account, which potentially creates conflicts with env prefix environments and automated testing.
- Internally it runs some basic prediction, which does not work well with sudden spikes of activity. And spikes in activity are very common for human-generated queries.

---

What I would suggest to use instead:

1. For realtime monitoring with custom fine-grained rules, use [SnowKill](https://github.com/littleK0i/SnowKill). Instead of just monitoring it would help you to proactively detect and terminate outlier SQL queries before excessive spend actually happens. This thing alone can massively increase visibility and reduce costs.

2. Create RESOURCE MONITORS both on account-level and warehouse level. Use daily granularity. Make sure to set both SUSPEND and NOTIFY triggers. Use `global_resource_monitor` in SnowDDL to attach monitors to warehouses. This would almost guarantee your account would not overspend even if something goes terribly wrong.

3. Ian from select.dev created some spending calculation [models](https://github.com/get-select/dbt-snowflake-monitoring) which help to allocate credits spend to specific queries. It includes idle time when warehouse waits for auto-suspend. If you combine this with aggregation by QUERY_HASH, you will be able to detect short and frequent queries. These queries may cause a lot of spend and hard to detect via other means.

---

There is still a problem with monitoring of serverless costs. But arguably it is not a big problem IF all serverless stuff is created only by SnowDDL and reviewed by human admin before deployment. Human users should not be able to create or trigger anything serverless. In this case you may use basic "Cost Management" tab in Snowflake GUI to keep an eye on it.

---

In theory, we may go with approach similar to "global" RESOURCE MONITORS and allow SnowDDL to attach objects to existing budgets. But budgets themselves should be created and managed outside of SnowDDL.

This is easier. But ... honestly, the feature is so flawed, I would highly recommend to ignore it altogether.

**Reply by onerishabh** [reply]
Thanks for sharing your thoughts and opinions here!
